icepyx/core/APIformatting.py

```diff
@@ -181,12 +184,56 @@ def to_string(params):
     return "&".join(param_list)
 
 
+ParameterType = Literal["CMR", "required", "subset"]
+# DevGoal: When Python 3.12 is minimum supported version, migrate to PEP695 style
+T = TypeVar("T", bound=ParameterType)
+
+
+class _FmtedKeysDescriptor:
+    """Enable the Parameters class' fmted_keys property to be typechecked correctly.
+
+    See: https://github.com/microsoft/pyright/issues/3071#issuecomment-1043978070
+    """
+
+    @overload
+    def __get__(  # type: ignore
+        self,
+        instance: 'Parameters[Literal["CMR"]]',
+        owner: Any,
+    ) -> CMRParams: ...
+
+    @overload
+    def __get__(
+        self,
+        instance: 'Parameters[Literal["required"]]',
+        owner: Any,
+    ) -> EGISpecificParams: ...
```

Comment:
the "required" type maps to either `EGISpecificParamsSearch` or `EGISpecificParamsDownload` but NOT ``EGISpecificParamsSubset` (all three are included in `EGISpecificParams`).

Reply:
Thank you for this clarity! :bow:

Reply:
Addressed here df78ca2  what do you think? I changed the names in a way that I feel is a bit clearer too.

Reply:
As I'm starting to look through these updates, I can't help but wonder why we're putting so much effort into perfecting type checking for code we're going to remove or substantially alter... given that the changes will be breaking anyway and as I learn more about when and how type checkers do (or don't!) actively "run" with code, it's not clear to me how having this in place helps us detect bugs for anything that's EGI specific.

Reply:
Woops, forgot to post here after we chatted on Zoom. To greatly simplify, since the conversation isn't fresh in my mind anymore, type annotations are really helpful to me in refactoring; as I remove code to replace it with something new, the typechecker tells me about how calling code expects to work with the new code I need to write. The errors sort of act as a guide during refactors of any size. I'm largely targeting the API interfaces here, but the same benefits apply to other refactors in these modules or code which uses them.

Also, IDEs (and I assume also LLMs?) will give better hints with type annotations!

Reply:
> Woops, forgot to post here after we chatted on Zoom.

No worries - I asked on Zoom because I thought GH ate my comment (I refreshed multiple times and couldn't find it, so I assumed it was lost in a draft box somewhere).